src/k8s/pkg/k8sd/controllers/node_configuration.go

```diff
@@ -0,0 +1,82 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/mitchellh/mapstructure"
+	v1 "k8s.io/api/core/v1"
+)
+
+type NodeConfigurationController struct {
+}
+
+func NewNodeConfigurationController() *NodeConfigurationController {
+	return &NodeConfigurationController{}
+}
+
+func (c *NodeConfigurationController) Run(ctx context.Context, snap snap.Snap) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+
+		client, err := k8s.NewClient(snap.KubernetesNodeRESTClientGetter("kube-system"))
+		if err != nil {
+			log.Println(fmt.Errorf("failed to create kubernetes node client: %w", err))
+			// This fails when the node is not bootstrapped or joined, acts like a retry timer
+			time.Sleep(3 * time.Second)
+			continue
+		}
+
+		reconcile := c.reconcile(ctx, snap)
+		client.WatchConfigMap(ctx, "kube-system", "k8sd-config", reconcile)
+
```

Comment:
then use a closure to capture the needed state here:

```suggestion
		client.WatchConfigMap(ctx, "kube-system", "k8sd-config", func(configMap *v1.ConfigMap) error { return c.reconcile(ctx, snap, configMap) })

```